feat: Add osnap plugin

daywiss opened the pull request:
### Summary

We have had great success offering oSnap as an optional alternative to Reality in the SafeSnap plugin. However, over time our needs have evolved. We have determined that to offer the best experience to our users, we need our own Snapshot plugin.

This PR seeks to add and integrate this new plugin.

We have intentionally not yet removed the legacy oSnap logic from the existing SafeSnap plugin, because we want users who have not migrated to still be able to use the old plugin as they have done up until now.

The new features / improvements of this plugin are:

Simplified transaction builder — since oSnap can execute an array of transactions and doesn't require them to be formatted as batches, we are able to greatly simplify the transaction builder UI.
Automatic module address detection — users no longer need to specify network and module addresses for safes in the plugin config.
Strongly typed data — we have types for all inputs and outputs that the plugin uses.
While we have wherever possible avoided making changes outside of the /plugins directory, there are some changes that exist outside:

in SpaceCreateVoting we have a use osnap checkbox (same as with the SafeSnap plugin)
in SettingsTreasuries we have an enable osnap button next to each treasury (only shown if the oSnap plugin is installed)


### How to test
[basic usage documentation](https://docs.google.com/document/d/1AXOVEztUdQwDwgzPAsCAXPGxPX3KG9bb2RAY5N8FCQY/edit?usp=sharing)

To test, we must run through the plugin's features and verify that they work on chain.

Enable / disable oSnap
Install the plugin on a space with Safe treasuries.
The activate osnap button should be visible.
Clicking the button should take the user to user Safe app.
Returning to Snapshot, the button for that treasury should now say oSnap active
Clicking that button again should take you to the Safe app again, where oSnap can be deactivated.

Transaction builder (Create)
Create a proposal and choose to use oSnap.
The final step should be the transaction builder.
The user should be able to add an arbitrary number of transactions.
The address fields and to fields should be required.
Once transactions are added and the proposal is submitted, the transactions should be persisted to ipfs.

Proposal
Once a proposal is created, the transactions should be visible there in a read-only state.
Once the vote is over:

if the vote failed, the user should not be able to do anything
if the vote passed, then the user should be able to submit the proposal with their own bond
if the proposal is submitted and has not been disputed in the Optimistic Oracle, the user should be able to request execution of the proposal
once execution is requested, the transactions should be completed on chain

### To-Do

- [ ]

<!--
### Self-review checklist
- [x] I have performed a full self-review of my changes
- [x] I have tested my changes on a preview deployment
- [x] I have tested my changes on different screen sizes (sm, md)
-->

### screens
![image](https://github.com/snapshot-labs/snapshot/assets/4429761/61de02a2-649b-4e18-a878-7849974ed805)
![image](https://github.com/snapshot-labs/snapshot/assets/4429761/7400b2e1-5cb8-4aa0-b7b9-b5f4a62c38d1)
![image](https://github.com/snapshot-labs/snapshot/assets/4429761/93624b6d-c1e5-4236-b705-a9c4e95954bd)
![image](https://github.com/snapshot-labs/snapshot/assets/4429761/0b21c32f-ab21-434c-83ee-a39e8625cb51)
![image](https://github.com/snapshot-labs/snapshot/assets/4429761/ed0b2d2c-c37b-4ee1-928f-2635c2961623)
![image](https://github.com/snapshot-labs/snapshot/assets/4429761/2cda260d-0a27-48e9-b829-10f52d2d3a5a)

samuveth commented:
Thanks! Going to review this soon. @bonustrack you want to also take a look?

bonustrack commented:
@daywiss I'm not able to login with the test deployment there seem to be a loop of request to Polygon node that end up with the error message "execution reverted", is this something you can reproduce?

Some suggestions based on the screenshot:

1: Space should be half smaller, 2: Space in blue should be removed, text should be aligned with green line. 3: There should be a space here same between previous section.
<img width="750" alt="image" src="https://github.com/snapshot-labs/snapshot/assets/16245250/c8ee4cb8-1fff-48e4-820a-d1cb00ec17c2">

Do you have a test space setup already?

daywiss commented:
> @daywiss I'm not able to login with the test deployment there seem to be a loop of request to Polygon node that end up with the error message "execution reverted", is this something you can reproduce?
> 
> Some suggestions based on the screenshot:
> 
> 1: Space should be half smaller, 2: Space in blue should be removed, text should be aligned with green line. 3: There should be a space here same between previous section. <img alt="image" width="750" src="https://user-images.githubusercontent.com/16245250/281144089-c8ee4cb8-1fff-48e4-820a-d1cb00ec17c2.png">
> 
> Do you have a test space setup already?

I will look into this issue with polygon. could you tell me what you are doing to get this error?

Ill take a look at making changes to the UI based on your suggestions

We have a test space here: https://snapshot-4w1pdwze1-snapshot.vercel.app/#/umadev.eth

daywiss commented:
@bonustrack I've updated the branch to incorporate latest changes on master, and noticed there were 2 issues.

1. login kept crashing, trying to index into an undefined var, so fixed that
2. after proposing, it wasnt detecting polygon network correctly due to comparing hex vs int ids, fixed that

Successfully installed osnap on polygon, made a proposal and it seemed fine after those changes, let me know

samuveth commented on src/components/SettingsPluginsBlock.vue:
```diff
@@ -14,6 +14,9 @@ const modalPluginsOpen = ref(false);
   if (props.isViewOnly) return;
+  // the oSnap plugin does not require any configuration
+  // so we don't need to open the modal
+  if (name === 'oSnap') return;
```
I think we should handle it like we do when selecting the plugin in the modal, when the plugin doesn't have `defaults` defined

https://github.com/snapshot-labs/snapshot/blob/335f6f2d59546023c96c17c4a763c8765fce02a9/src/components/ModalPlugins.vue#L32

daywiss replied:
could you go into more detail? i dont understand if you want me to remove the changes in the `handleEditPlugins` and instetad put it in the `selectPlugin` function instead, or move the default checking from `selectPlugin` to `handleEditPlugins` function

samuveth commented on src/plugins/oSnap/components/Input/Amount.vue:
```diff
@@ -0,0 +1,47 @@
+</script>
+
+<template>
+  <UiInput
```
Would be great if you could use Tune components instead of `/Ui` as we are in the process of deprecating these. But if this will create major difficulties for you(e.g making many changes to Tune) then I;d say we can refactor at a later time.

samuveth commented on src/plugins/oSnap/plugin.json:
```diff
@@ -0,0 +1,12 @@
+  "author": "UMA",
+  "website": "https://safe.global",
+  "icon": "ipfs://QmQjZaheMTLRrk22mrGCkGk2LNoNqqYMR8Bxtwa8mBHyc9",
+  "defaults": {
```
If the Osnap plugin doesn't require settings it would be better to add these values elsewhere. As having this here will open the JSON field when selecting the plugin.

daywiss replied:
i see, ok, i guess this is related to your other comment!

daywiss replied:
removing the defaults still allows the plugin to be clicked and modal open to add custom json. that seems unexpected

samuveth replied:
Yes, this is a bug, we should also disable the edit modal when there are no defaults

samuveth commented:
@daywiss can you add me ass an author? 0x3901D0fDe202aF1427216b79f5243f8A022d68cf

daywiss commented:
> 0x3901D0fDe202aF1427216b79f5243f8A022d68cf

@samuveth added you as an author to https://snapshot.org/#/umadev.eth/

samuveth reviewed (approved):
tACK

bonustrack commented:
@daywiss I've noticed some issues:

- A typo, "collectable" instead of "collectible", I would rather rename the actions to "Send token" and "Send NFT", this should be more obvious for users:
<img width="427" alt="image" src="https://github.com/snapshot-labs/snapshot/assets/16245250/25f3c00e-0d79-496d-9f55-b228d59173bd">

- We should use "Sentence case" and not "Title Case" everywhere in the app, I found some place in the integration where it's not the case:
<img width="585" alt="image" src="https://github.com/snapshot-labs/snapshot/assets/16245250/558b05ca-53b2-4e74-8726-e17706014026">
<img width="230" alt="image" src="https://github.com/snapshot-labs/snapshot/assets/16245250/276f1541-4857-4c59-b2c9-653a86e5234b">
<img width="158" alt="image" src="https://github.com/snapshot-labs/snapshot/assets/16245250/3893c341-adab-40fe-9d2b-88b6c6ae16e8">

- In the modal to activate oSnap the text in footer is not visible in dark mode:
<img width="551" alt="image" src="https://github.com/snapshot-labs/snapshot/assets/16245250/2f7d7819-12fd-4306-b1b8-e40ab61c2b05">

- Any reason to show the data of the tx? I feel it's not obvious what the tx is going to do even for a simple action like sending a token:
<img width="660" alt="image" src="https://github.com/snapshot-labs/snapshot/assets/16245250/39b40264-fab2-4e46-850d-902c403e030e">

It would be great if we can hop on a call and go through the flow together so I can suggest some last cosmetic change.

Lizdaldal commented:
Hi @bonustrack, thanks for finding these issues! We'll get these fixed asap.

@daywiss is out this week for Thanksgiving in The States. Any chance you're available tomorrow to jump on a call with me and @gsteenkamp89 to go through the flow? We're available at 3pm or 4pm UTC. 

If you're not available tomorrow, @daywiss is back on Monday and is available 1pm-7pm UTC. Let me know what works best for you.

bonustrack commented:
@Lizd 3pm today would work for me, we can sync in our Telegram group

samuveth commented on src/components/SettingsTreasuryActivateOsnapButton.vue:
```diff
@@ -1,24 +1,58 @@
+const activeStyles = computed(() => {
+  return theme.value === 'light'
+    ? {
+        div: 'bg-[hsla(122,100%,45%,0.13)] text-[hsla(122,100%,21%,1)]',
```
Using custom colors with not work for our skins on Snapshot. We don't only support light and dark mode, but also custom skins for custom domains, you can test this by changing the key for a skin in domains.json to `localhost`. Please use only available color variable throughout your whole pr.

Custom colors won't work for our skins on Snapshot. We support not only light and dark mode but also custom skins for custom domains. You can test this by changing a key in `domains.json` to `localhost`. 

Please try to use only the available color variables throughout your entire PR.

samuveth replied:
There is also a bug here with `Activated` button on light mode. 
<img width="781" alt="image" src="https://github.com/snapshot-labs/snapshot/assets/51686767/4eabae7f-49c2-4977-9157-a702cf765d7f">
